### DISCUSSION TITLE
Using matrix to specify shell: is it possible?

### DISCUSSION THREAD

**Discussion by mlocati** [community/community#25544] [Actions]
<p>I’m trying to run a job with two different shells (powershell and pwsh).</p>
<p><strong>Attempt <span class="hashtag">#1</span></strong></p>
<p>If I specify this workflow configuration</p>
<pre><code>name: Test 1
on:
  pull_request:
  push:
    branches:
      - master
jobs:
  test:
    runs-on: windows-latest
    strategy:
      matrix:
        shell: [powershell, pwsh]
    steps:
      - name: Print shell version
        run: $PSVersionTable
</code></pre>
<p>The output for both the executions is the same (like if matrix.shell is not taken in account):</p>
<pre><code>Run $PSVersionTable
    $PSVersionTable
    shell: C:\Program Files\PowerShell\6\pwsh.EXE -command ". '{0}'"

Name Value
---- -----
PSVersion 6.2.3
PSEdition Core
GitCommitId 6.2.3
OS Microsoft Windows 10.0.17763
Platform Win32NT
PSCompatibleVersions {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion 2.3
SerializationVersion 1.1.0.1
WSManStackVersion
</code></pre>
<p><strong>Attempt <span class="hashtag">#2</span></strong></p>
<p>If I specify this workflow configuration</p>
<pre><code>name: Test 2
on:
  pull_request:
  push:
    branches:
      - master
jobs:
  test:
    runs-on: windows-latest
    strategy:
      matrix:
        shell_name: [powershell, pwsh]
    steps:
      - name: Print shell version
        shell: ${{ matrix.shell_name }}
        run: $PSVersionTable
</code></pre>
<p>The job isn’t executed, because of this error:</p>
<pre><code>### ERRORED 07:43:52Z

- Your workflow file was invalid: The pipeline is not valid. .github/workflows/test2.yml (Line: 15, Col: 16): Unrecognized named-value: 'matrix'. Located at position 1 within expression: matrix.shell_name
</code></pre>
<p>Is there a way to use matrix to specify the shell to be used?</p>

**Comment by BrightRan** [comment, marked as answer]
<p>I’m afraid matrix job does not support to set shell as configuration options for a GitHub virtual environment. On Windows, it always uses pwsh to run the commands by default, if you do not specify the shell on the steps.</p>

**Reply by beamerblvd** [reply]
Well that's really unfortunate, because `pwsh` does not fail when commands fail [per this unfixed bug](https://github.com/actions/runner-images/issues/6668), so I can't make just my windows runners use `powershell` without having a completely different workflow script. :-(

**Comment by mlocati** [comment]
<p>Thanks for the quick reply, <a class="mention" href="/brightran">@brightran</a>!</p>
<p>I wanted to test my PowerShell module both on PowerShell and on PowerShell Core, and using matrix would have lead to a smaller workflow file.</p>
<p>Not a huge problem, though.</p>

**Comment by gmeligio** [comment]
I found this problem now when testing cross-platform commands between Linux bash and Windows cmd, and unfortunately, it's not yet supported. Here is a related issue https://github.com/actions/runner/issues/444
